### PR TITLE
Fulltext search optimize

### PR DESCRIPTION
**Pull request by maggarwal13**

### CODE REVIEW

File: whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/AutocompleteSearch.java [hunk 1]
```diff
@@ -52,9 +56,12 @@
 
     private final FullTextIndex fullTextIndex;
 
+    private final RpslObjectDao objectDao;
+
     @Autowired
-    public AutocompleteSearch(final FullTextIndex fullTextIndex) {
+    public AutocompleteSearch(final FullTextIndex fullTextIndex, @Qualifier("jdbcRpslObjectSlaveDao") final RpslObjectDao objectDao) {
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/AutocompleteSearch.java:62]: Rename to rpslObjectDao for consistency

File: whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/AutocompleteSearch.java [hunk 2]
```diff
@@ -88,17 +95,22 @@ public AutocompleteSearch(final FullTextIndex fullTextIndex) {
                 for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
                     final Document doc = indexSearcher.doc(scoreDoc.doc);
                     final Map<String, Object> result = Maps.newLinkedHashMap();
-                    result.put("key", doc.get(FullTextIndex.LOOKUP_KEY_FIELD_NAME));
-                    result.put("type", doc.get(FullTextIndex.OBJECT_TYPE_FIELD_NAME));
+
+                    final RpslObject rpslObject =   objectDao.getByKey(
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/AutocompleteSearch.java:99]: Handle EmptyResultDataAccessException here (MariaDB and Lucene may be out of sync)

File: whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/FullTextSearch.java [hunk 3]
```diff
@@ -69,15 +75,18 @@
     private final AccessControlListManager accessControlListManager;
     private final Source source;
     private final Version version;
+    private final RpslObjectDao objectDao;
 
     @Autowired
     public FullTextSearch(final FullTextIndex fullTextIndex,
+                          @Qualifier("jdbcRpslObjectSlaveDao") final RpslObjectDao objectDao,
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/FullTextSearch.java:82]: Rename to rpslObjectDao for consistency

File: whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/FullTextSearch.java [hunk 4]
```diff
@@ -164,23 +173,24 @@ protected SearchResponse doSearch(final IndexReader indexReader, final TaxonomyR
 
                     indexSearcher.search(query, MultiCollector.wrap(topFieldCollector, facetsCollector));
 
-                    final List<Document> documents = Lists.newArrayList();
+                    final Map<RpslObject, Document> rpslObjectToDocument = Maps.newHashMap();
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/FullTextSearch.java:176]: I would flip this Map around (document with a matching RpslObject) or create a list of "pairs"

File: whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/IndexTemplate.java [hunk 5]
```diff
@@ -229,24 +229,17 @@ protected void account(final RpslObject rpslObject) {
...
+                        return  Integer.parseInt(field.stringValue());
                     }
-                } else {
-                    attributes.add(new RpslAttribute(AttributeType.getByName(field.name()), field.stringValue()));
                 }
             }
 
-            attributes.add(new RpslAttribute(AttributeType.SOURCE, source.getName()));
-            return new RpslObject(objectId, attributes);
+           return 0;
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/IndexTemplate.java:241]: If this should never happen, then Throw IllegalStateException

File: whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/FullTextIndex.java [hunk 6]
```diff
@@ -289,30 +293,54 @@ private void addEntry(final IndexWriter indexWriter, final TaxonomyWriter taxono
...
+                continue;
+            }
+
+            attributes.add(new RpslAttribute(attribute.getKey(), filterRpslAttribute(attribute.getType(), attribute.getValue())));
+        }
+
+        return new RpslObject(rpslObject.getObjectId(), attributes);
+    }
+
+    //need to filter as list (not set) as two AUTH MD5 after filter will return 1 entry in set
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/FullTextIndex.java:320]: Remove this commented block altogether?

File: whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/FullTextSearch.java (hunk 4 above)
Comment (maggarwal13) [reply whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/FullTextSearch.java:176]: I would rely on Rpslobject as key  rather then luecene document. As Rpslobject have our own implementation of  equals and hashcode

Comment (eshryane) [reply whois-api/src/main/java/net/ripe/db/whois/api/fulltextsearch/FullTextSearch.java:176]: Good point, thanks!

**Pull request merged**